Utilities/XrdAdaptor/src/XrdSource.cc

```diff
@@ -76,6 +76,110 @@ class DelayedClose : boost::noncopyable, public XrdCl::ResponseHandler
+ * invoked from an XrdCl callback (that is, we don't need an available callback
+ * thread to timeout).
+ */
+class QueryAttrHandler : boost::noncopyable, public XrdCl::ResponseHandler
+{
+public:
+
+    virtual ~QueryAttrHandler() {}
+
+
+    static XrdCl::XRootDStatus query(XrdCl::FileSystem &fs, const std::string &attr, std::chrono::milliseconds timeout, std::string &result)
+    {
+        std::shared_ptr<QueryAttrHandler> handler(new QueryAttrHandler());
+        XrdCl::Buffer arg(attr.size());
+        arg.FromString(attr);
+
+        // On error or exception thrown, drop the reference to ourself
+        std::unique_ptr<QueryAttrHandler, std::function<void(QueryAttrHandler*)>> self_ref_guard(nullptr, [&](QueryAttrHandler*) {handler->m_self.reset();});
+        handler->m_self = handler;
+
+        XrdCl::XRootDStatus st = fs.Query(XrdCl::QueryCode::Config, arg, handler.get());
+        if (!st.IsOK())
+        {
+            return st;
+        }
+
+        // Successfully registered the callback; keep the self-reference until the
+        // callback release it.
+        self_ref_guard.release();
+
+        std::unique_lock<std::mutex> guard(handler->m_mutex);
+        // Wait until some status is available or a timeout.
+        handler->m_condvar.wait_for(guard, timeout, [&]{return handler->m_status.get();});
+
+        if (handler->m_status)
+        {
+            if (handler->m_status->IsOK())
+            {
+                result = handler->m_response->ToString();
+            }
+            return *(handler->m_status);
+        }
+        else
+        {   // We had a timeout; construct a reasonable message.
+            return XrdCl::XRootDStatus(XrdCl::stError, XrdCl::errSocketTimeout, 1, "Timeout when waiting for query callback.");
+        }
+    }
+
+
+private:
+
+    QueryAttrHandler() {}
+
+
+    virtual void HandleResponse(XrdCl::XRootDStatus *status, XrdCl::AnyObject *response ) override
+    {
+        // NOTE: we own the status and response pointers.
+        std::unique_ptr<XrdCl::AnyObject> response_mgr;
+        response_mgr.reset(response);
+        // On exit from this function, release the self-reference.  Order matters!  m_self reset must be done after
+        // condvar is notified.
+        std::unique_ptr<char, std::function<void(char*)>> self_ref_guard(nullptr, [&](char *) {m_self.reset();});
+        // On function exit, notify any waiting threads.
+        std::unique_ptr<char, std::function<void(char*)>> notify_guard(nullptr, [&](char *) {m_condvar.notify_all();});
+
+        {
+            // m_mutex protects m_status
+            std::unique_lock<std::mutex> guard(m_mutex);
+            // On exit from the block, make sure m_status is set; it needs to be set before we notify threads.
+            std::unique_ptr<char, std::function<void(char*)>> exit_guard(nullptr, [&](char *) {m_status.reset(new XrdCl::XRootDStatus(XrdCl::stError, XrdCl::errInternal));});
```

Comment:
Yup, it does!

Since I want to `release()` the data later, I felt it was appropriate here.